huggingface/accelerate
# Add Upcasting for FSDP in Mixed Precision. Add Concept Guide for FSPD and DeepSpeed.

# What does this PR do?

This PR address the issues identified in #2624, whereby differences were found between how DeepSpeed (DS) and FSDP handle the sharded parameters during mixed precision. To address this we have:
- added logic in `accelerate.prepare` to upcast the FSDP sharded paramaters, if it has been detected that:
   1. mixed precision has been activated, and,
   2. the sharded parameters are in low precision.
- added concept guide to clarify to how DS and FSDP handles mixed precision. Also to inform that 🤗  Accelerate will upcast FSDP sharded weights.
- added warning to inform user, that if the weights were upcasted, then this will affect the precision in which the checkpoint will be saved. 

In addition to the above, we also:
- int the above mentioned concept guide, clarified the equivalences between DS and FSDP configs, enabling users to better transition between DS/FSDP.
- updated the outdated [accelerate launch cli commands](https://huggingface.co/docs/accelerate/en/package_reference/cli#accelerate-launch).

Checklist:
- [x] test with low mem 
- [x] test with `SHARD_GRAD_OP`
- [x] test with CPU offload
- ~consider the [impacts of fp8 training](https://github.com/huggingface/accelerate/issues/2624#issuecomment-2050404125) suggested by @stas00.~ Update: this has to be done later 

To test and reproduce:
-  get the FSDP/ DS accelerate configurations, and the reproduction script (call it `learning_rate_repro.py`)  [from here](https://github.com/huggingface/accelerate/issues/2624#issue-2225498257).  
- run the script with `accelerate_fspd.yaml` config and without `--bf16` flag; this gets the `bf16` case in the below plot. The script will load the model in `bfloat16` (since `load_model_dtype` was default) and **turn off mixed precision**.
```
    accelerate launch \
        --num_processes 4 \
        --config_file accelerate_fsdp.yaml \
        learning_rate_repro.py  \
          --num_train_epochs 10 \
          --output_dir './results' \
          --per_device_train_batch_size 10 \
          --lr_scheduler_type "linear" \
          --learning_rate 1e-6 \
          --logging_steps 1 \
          --save_strategy 'no' 
```
- run the same adding `--bf16` (this gets `bf16-with-mp` in the plot below). This runs `bfloat16` model in FSDP **with mixed precision**.
- run the same benchmark with a DS `accelerate_ds.yaml` config (here `--bf16` is irrelevant, DS will handle both cases similarly). 

In the plot below, we can see that with the casting logic put in:
- if we detect that the model is in low precision `bfloat16` but mixed precision is turned on (`--bf16`), then we perform the upcasting. Hence we see that the loss curves of DS and FSDP will now be the same with the added casting logic.
- *If the logic was absent*, then we would have observed that FSDP would not converge, regardless if `--bf16` had been supplied or not

**Sample of New Warnings**

We have added some logic to reduce repetitive warnings: 

```
/data/flim/accelerate/src/accelerate/accelerator.py:1492: UserWarning: Upcasting low precision parameters in MistralForCausalLM, mixed precision is turned on: model.embed_tokens.weight,model.norm.weight,lm_head.weight.
  warnings.warn(
/data/flim/accelerate/src/accelerate/accelerator.py:1492: UserWarning: Upcasting low precision parameters in MistralDecoderLayer, mixed precision is turned on: self_attn.q_proj.weight,self_attn.k_proj.weight,self_attn.v_proj.weight,self_attn.o_proj.weight,mlp.gate_proj.weight,mlp.up_proj.weight,mlp.down_proj.weight,input_layernorm.weight,post_attention_layernorm.weight.
  warnings.warn(
```



**Performance of low precision models under FSDP and DS and mixed precision, with the proposed casting fix** 

This was plotted when comparing FSDP (Full Shard) and DeepSpeed (Zero3)
![image](https://github.com/huggingface/accelerate/assets/8325951/15addc48-d8f6-4bfd-8acb-52305d57415c)

This was plotted when comparing FSDP in various modes, namely GradOp and CPUOffload (while full-sharding)
![image](https://github.com/huggingface/accelerate/assets/8325951/c059ec0d-8659-478c-af2f-18dda83b2354)


## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
      to it if that's the case.
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).


## Who can review?

Tagging @stas00, @muellerzr and @pacman100  first. We can add more reviewers later.
